File: awa/sdk_dist/csharp/Awa.Client/Utilities/Workflow/ReadFileAndParseWorkflow.cs
using System;
using System.Threading.Tasks;
using Awa.Client.Constants;
using Awa.Client.Models;

namespace Awa.Client.Utilities.Workflow
{
    public static partial class WorkflowUtilities
    {
        /// <summary>
        /// Read and parse a file from a specified path, converting to markdown if supported.
        ///
        /// Uses MarkItDown to parse specific whitelisted document formats:
        /// - Documents: .pdf, .docx, .pptx, .xlsx, .xls
        /// - Web: .html, .htm
        /// - Data: .csv
        /// - E-books: .epub
        /// - Email: .msg (Outlook messages)
        ///
        /// All other file types (including .txt, .md, .py, etc.) are returned as-is without parsing
        /// when strict=False, or raise an exception when strict=True.
        /// </summary>
        /// <param name="filePath">The file path to read and parse</param>
        /// <param name="defaultValue">A string to return if the file does not exist</param>
        /// <param name="strict">If True, raise exception for unsupported file types; if False, return raw content</param>
        /// <returns>
        /// A string containing the parsed content (markdown) for supported formats,
        /// or raw content for unsupported formats when strict=False.
        /// </returns>
        /// <exception cref="Temporalio.Exceptions.ApplicationFailureException">When strict=True and file type is not supported</exception>
        public static async Task<string?> ReadFileAndParseWorkflow(
            string filePath,
            string? defaultValue = null,
            bool strict = false)
        {
            var input = new ReadFileAndParseInput
            {
                FilePath = filePath,
                DefaultContent = defaultValue,
                Strict = strict
            };

            var result = await Temporalio.Workflows.Workflow.ExecuteChildWorkflowAsync<string?>(
                AwaConstants.WorkflowReadFileAndParse,
                new object?[] { input },
                new()
                {
                    TaskQueue = AwaConstants.AwaDefaultTaskQueue
                });

            return result;
        }
    }
}

File: awa/awa/sdk/utils/workflow/read_file_and_parse_workflow.py
from pathlib import Path

from temporalio import workflow

from awa.sdk import constants
from awa.sdk.models.read_file_and_parse_input import ReadFileAndParseInput


async def read_file_and_parse_workflow(
    file_path: str | Path,
    default: str | None = None,
    strict: bool = False,
) -> str:
    """Read and parse a file from a specified path, converting to markdown if supported.

    Uses MarkItDown to parse specific whitelisted document formats:
    - Documents: .pdf, .docx, .pptx, .xlsx, .xls
    - Web: .html, .htm
    - Data: .csv
    - E-books: .epub
    - Email: .msg (Outlook messages)

    All other file types (including .txt, .md, .py, etc.) are returned as-is without parsing
    when strict=False, or raise an exception when strict=True.

    Args:
        file_path: The file path to read and parse.
        default: A string to return if the file does not exist.
        strict: If True, raise exception for unsupported file types; if False, return raw content.

    Returns:
        A string containing the parsed content (markdown) for supported formats,
        or raw content for unsupported formats when strict=False.

    Raises:
        InvalidInputApplicationError: When strict=True and file type is not supported.

    """
    result = await workflow.execute_child_workflow(
        workflow=constants.WORKFLOW_READ_FILE_AND_PARSE,
        arg=ReadFileAndParseInput(
            file_path=str(file_path),
            default_content=default,
            strict=strict,
        ),
        task_queue=constants.AWA_DEFAULT_TASK_QUEUE,
    )
    return result
